### PR TITLE
Avoid unnecessary dupe transcode jobs, retry on gdrive->s3 upload errors a few times before raising

### LINKED ISSUE #1215: Retries on syncing gdrive files can lead to dupe transcodes, overwritten VideoFile.destination_ids
### Steps to Reproduce

It's possible that multiple tasks to process a `DriveFile` could occur - for example if the "Sync w/GDrive" button is pressed while the regularly scheduled task `import_recent_files` is running.

### Expected Behavior
Despite any dupe processing tasks, only one transcode job should be created per video. 

### Actual Behavior

More than one transcode job is created, and when the second webhook request comes in from AWS, it overwrites the `VideoFile.destination_id` value set by the first webhook.

### PR DESCRIPTION
**Pull request by mbertrand**
#### Pre-Flight checklist
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes #1215 

#### What's this PR do?
- Adds additional DriveFile statuses under which processing for existing objects should be skipped if the `name` and `checksum` values have not changed.
- Prevents another transcode job from being created if there is a prior transcode job that didn't complete/fail.
- If another transcode job does send a webhook back to studio, update the affected `VideoFile.status` back to `Created` so it will be picked up by the `videos.tasks.upload_youtube_videos` task.
- Adds the `@retry_on_fail` decorator to the `stream_to_s3` function in case there is[ an occasional failure streaming from google drive to S3](https://sentry.io/organizations/mit-office-of-digital-learning/issues/3171662861/?environment=production&project=5739976).

#### How should this be manually tested?
- Copy `DRIVE_*` and `AWS_*` settings from RC
- Upload a small video and non-video file to a site's google drive folder, under the appropriate subfolder for each (`videos_final`, `files_final`).
- Click the 'Sync w/Google Drive' button, wait for the new resources to show up.
- Verify that you can link to or embed the non-video file.
- Go to http://localhost:8043/admin/videos/video/, there should be a new object for the video you uploaded, and it should have one `VideoJob` (at the bottom).
- Using Postman or curl, make a POST request to http://localhost:8043/api/transcode-jobs/ with the following data, subsituting the correct values including the `VideoJob.job_id`:
```json
{
  "version": "0",
  "id": "c120fe11-87db-c292-b3e5-1cc90740f6e1",
  "detail-type": "MediaConvert Job State Change",
  "source": "aws.mediaconvert",
  "account": "<AWS_ACCOUNT_ID>",
  "time": "2021-08-05T16:52:33Z",
  "region": "us-east-1",
  "resources": [
    "arn:aws:mediaconvert:us-east-1:<AWS_ACCOUNT_ID>:jobs/<VIDEOJOB_ID>"
  ],
  "detail": {
    "timestamp": 1628172900136,
    "accountId": "<AWS_ACCOUNT_ID>",
    "queue": "arn:aws:mediaconvert:us-east-1:<AWS_ACCOUNT_ID>:queues/Default",
    "jobId": "<VIDEOJOB_ID>",
    "status": "COMPLETE",
    "userMetadata": {},
    "outputGroupDetails": [
      {
        "outputDetails": [
          {
            "outputFilePaths": [
              "s3://<AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<SITE_NAME>/<DRIVE_FILE_ID>/<VIDEO_FILENAME>.mp4"
            ],
            "durationInMs": 132033,
            "videoDetails": {
              "widthInPx": 1280,
              "heightInPx": 720
            }
          },
          {
            "outputFilePaths": [
              "s3://<AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<SITE_NAME>/<DRIVE_FILE_ID>/<VIDEO_FILENAME>_360p_16_9.mp4"
            ],
            "durationInMs": 132033,
            "videoDetails": {
              "widthInPx": 640,
              "heightInPx": 360
            }
          },
          {
            "outputFilePaths": [
              "s3://<AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<SITE_NAME>/<DRIVE_FILE_ID>/<VIDEO_FILENAME>_360p_4_3.mp4"
            ],
            "durationInMs": 132033,
            "videoDetails": {
              "widthInPx": 480,
              "heightInPx": 360
            }
          }
        ],
        "type": "FILE_GROUP"
      }
    ]
  }
}
```
- Verify there are now 3 `VideoFile` objects for the video, all with blank `destination_id`s and `status=Created`
- In a shell, run the following:
```python
from gdrive_sync.api import *
video = Video.objects.order_by("-created_on").first()
drive_file = DriveFile.objects.get(video=video)
transcode_gdrive_video(drive_file)
```
- Reload the admin page, verify there is still only one `VideoJob` 
- Go back to the shell and run this:
```python
VideoJob.objects.filter(job_id="<your job id>").update(status=VideoJobStatus.FAILED)
transcode_gdrive_video(drive_file)
```
- Reload the admin page, there should now be two `VideoJob` objects
- Update the Youtube `VideoFile` with a non-blank `destination_id` value, `destination_status=processed`, and `status=Complete` and save the changes.
- Run the POST request from above again, this time with the newest `VideoJob.job_id`
- Verify that the 3 `VideoFile` objects for the video once again have all blank `destination_id`s and `status=Created`

### CODE REVIEW

File: gdrive_sync/api.py [hunk 1]
```diff
@@ -189,21 +197,10 @@ def process_file_result(
...
-        service = get_drive_service()
-        permission = (
-            service.permissions()
-            .create(
-                supportsAllDrives=True,
-                body={"role": "reader", "type": "anyone"},
-                fileId=drive_file.file_id,
-            )
-            .execute()
-        )
```
Comment (mbertrand) [inline gdrive_sync/api.py:206]: This code is no longer necessary due to PR #1169

File: gdrive_sync/api.py [hunk 2]
```diff
@@ -227,13 +224,6 @@ def stream_to_s3(drive_file: DriveFile):
         )
         drive_file.update_status(DriveFileStatus.UPLOAD_FAILED)
         raise
-    finally:
-        if service and permission:
-            service.permissions().delete(
-                supportsAllDrives=True,
-                permissionId=permission["id"],
-                fileId=drive_file.file_id,
-            ).execute()
```
Comment (mbertrand) [inline gdrive_sync/api.py:236]: Ditto

File: gdrive_sync/api_test.py [hunk 3]
```diff
@@ -362,6 +372,60 @@ def test_process_file_result(
         assert drive_file.checksum == checksum
 
 
+@pytest.mark.parametrize("status", [DriveFileStatus.UPLOADING, DriveFileStatus.FAILED])
+@pytest.mark.parametrize("same_checksum", [True, False])
+@pytest.mark.parametrize("same_name", [True, False])
+def test_process_file_result_update(settings, mocker, status, same_checksum, same_name):
+    """
+    An existing drive file should not be processed again is the checksum and name are the same, ond the status
```
Comment (ChristopherChudzicki) [inline gdrive_sync/api_test.py:380]:
```suggestion
    An existing drive file should not be processed again if the checksum and name are the same, ond the status
```

**Review by ChristopherChudzicki — APPROVED** [review]
👍 Does what you said! I'm particularly familiar with this area yet, but this makes sense.

Noticed one minor typo.

**Pull request merged**